pkg/status/peers.go

```diff
@@ -0,0 +1,312 @@
+// Copyright (c) 2016 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package status
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"regexp"
+	"strings"
+	"time"
+
+	"reflect"
+
+	apiv3 "github.com/projectcalico/api/pkg/apis/projectcalico/v3"
+
+	"github.com/olekukonko/tablewriter"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// Check for Word_<IP> where every octate is seperated by "_", regardless of IP protocols
+// Example match: "Mesh_192_168_56_101" or "Mesh_fd80_24e2_f998_72d7__2"
+var bgpPeerRegex = regexp.MustCompile(`^(Global|Node|Mesh)_(.+)$`)
+
+// Mapping the BIRD/GoBGP type extracted from the peer name to the display type.
+var bgpTypeMap = map[string]apiv3.BGPPeerType{
+	"Global": apiv3.BGPPeerTypeGlobalPeer,
+	"Mesh":   apiv3.BGPPeerTypeNodeMesh,
+	"Node":   apiv3.BGPPeerTypeNodePeer,
+}
+
+// Expected BIRD protocol table columns
+var birdExpectedHeadings = []string{"name", "proto", "table", "state", "since", "info"}
+
+// bgpPeer is a structure containing details about a BGP peer.
+type bgpPeer struct {
+	peerIP   string
+	peerType string
+	state    string
+	since    string
+	bgpState string
+	info     string
+}
+
+func (b *bgpPeer) toNodeStatusAPI() apiv3.CalicoNodePeer {
+	info := b.bgpState
+	if b.info != "" {
+		info += " " + b.info
+	}
+
+	return apiv3.CalicoNodePeer{
+		PeerIP: b.peerIP,
+		Type:   bgpTypeMap[b.peerType],
+		State:  b.state,
+		Since:  b.since,
+		Reason: info,
+	}
+}
+
+// Unmarshal a peer from a line in the BIRD protocol output.  Returns true if
+// successful, false otherwise.
+func (b *bgpPeer) unmarshalBIRD(line, ipSep string) bool {
```

Comment:
Any reason to use a `bool` here rather than an `error`? 

It seems to be that we should always expect this to pass unless there is a bug in our code or a real error in BIRD generating bad output, and we should be alerting the user to that.

I think as it is right now, we just log at debug level for errors, which might not be enough.

Reply:
This is inherited from calicoctl. I think the idea is we could skip some lines which does not provide any information. It is not an error, it just means we don't need to worry about those lines. 

I'll update the code to make warning level for errors we do care about.